### ISSUE TITLE
Python 3.11 loses the ability to set PYTHON_DECIMAL_WITH_MACHINE

### ISSUE THREAD

**Issue by Bo98** [python/cpython#98557]
# Bug report

Python 3.10 had the ability to set `PYTHON_DECIMAL_WITH_MACHINE` to override the choice of configuration for the `_decimal` module:
https://github.com/python/cpython/blob/dcb342b5f9d931b030ca310bf3e175bbc54df5aa/setup.py#L2388-L2397

Since Python 3.11, this is no longer possible. This feature was necessary, on macOS particularly, with the `--with-system-libmpdec` option if that system libmpdec is configured differently to the default Python config. On macOS, the default Python config forces universal, while setting `PYTHON_DECIMAL_WITH_MACHINE` allowed it to be single-arch.

`libmpdec` produces different headers depending on how it was built, which is why the setting is important to be able to override. Without it, the `_decimal` module will fail to compile if the default does not match how system `libmpdec` was built.

Homebrew's Python currently depends on this feature.

A test within CPython also seems to depend on this feature: https://github.com/python/cpython/blob/f4c03484da59049eb62a9bf7777b963e2267d187/Modules/_decimal/tests/runall-memorydebugger.sh#L63-L79

# Your environment

- CPython versions tested on: 3.11.0rc2
- Operating system and architecture: macOS 12 (x86_64 and arm64)

**Comment by AlexWaygood** [comment]
Cc. @rhettinger for `decimal` module expertise.

**Comment by erlend-aasland** [comment]
FTR, this was removed in GH-29541 (cc. @tiran)

**Comment by erlend-aasland** [comment]
AFAICS from GH-29541, it should be possible to set `PYTHON_DECIMAL_WITH_MACHINE` via CFLAGS (for example `LIBMPDEC_CFLAGS`). @Bo98, can you please test this?

**Comment by Bo98** [comment]
What reads `PYTHON_DECIMAL_WITH_MACHINE`? It's not a define to be passed to C but something that used to change what item is selected from: https://github.com/python/cpython/blob/75a6fadf369315b27e12f670e6295cf2c2cf7d7e/configure.ac#L3882-L3891

Unless you suggest that I pass `LIBMPDEC_CFLAGS=-DCONFIG_64=1 -DASM=1` instead on x86_64 and `LIBMPDEC_CFLAGS=-DCONFIG_64=1 -DANSI=1 -DHAVE_UINT128_T=1` on arm64? (Assuming this list is never changing.)

**Comment by erlend-aasland** [comment]
Ah, sorry, you are right. I misread the diff from GH-29541; it is used as an index, not a CFLAG.

**Comment by erlend-aasland** [comment]
You mentioned Homebrew? Does this prevent Homebrew from building and distributing Python? Is there a Homebrew ticket for this?

**Comment by Bo98** [comment]
> Is there a Homebrew ticket for this?

Not yet considering we don't ship betas, but what we'll probably do for 3.11.0 (and what I've done in a local test rc2 build) is a `s/libmpdec_machine=universal/libmpdec_machine=x64/` (`uint128` on arm64) on `configure`. Not a desirable long-term solution but should be ok for a short-term fix.

> I suggest you use LIBMPDEC_CFLAGS to set these flags depending on how your libmpdec was built. That seems to me to be a more robust approach than letting configure chose flags which might not reflect how your libmpdec was built.

`PYTHON_DECIMAL_WITH_MACHINE` did directly match the option you would pass to libmpdec configure. If you passed `MACHINE=x64` to libmpdec while building then you would pass the same to Python.

It made sense and is what Stefan Krah recommended we did.

Might I propose an alternative way forward here however. I propose looking into not setting any CFLAGS for the system libmpdec and checking `MPD_CONFIG_64` and `MPD_CONFIG_32` in `_decimal` source code, given mpdecimal.h should tell you how it was built by setting those.

FWIW, this is what the relevent section of mpdecimal.h looks like on a different configurations of system libmpdec:

<details>
<summary>Universal</summary>

```
/* Mac OS X: support for building universal binaries */
#if defined(MPD_CONFIG_64) || defined(MPD_CONFIG_32)
  #error \"cannot use MPD_CONFIG_64 or MPD_CONFIG_32 with universal header.\"
#endif

#if defined(CONFIG_64) || defined(CONFIG_32)
  #error \"cannot use CONFIG_64 or CONFIG_32 with universal header.\"
#endif

#if defined(__ppc__)
  #define MPD_CONFIG_32 1
  #ifdef UNIVERSAL
    #define CONFIG_32
    #define ANSI
  #endif
#elif defined(__ppc64__)
  #define MPD_CONFIG_64 1
  #ifdef UNIVERSAL
    #define CONFIG_64
    #define ANSI
  #endif
#elif defined(__i386__)
  #define MPD_CONFIG_32 1
  #ifdef UNIVERSAL
    #define CONFIG_32
    #define ANSI
  #endif
#elif defined(__x86_64__)
  #define MPD_CONFIG_64 1
  #ifdef UNIVERSAL
    #define CONFIG_64
    #define ASM
  #endif
#elif defined(__arm64__)
  #define MPD_CONFIG_64 1
  #ifdef UNIVERSAL
    #define CONFIG_64
    #define ANSI
  #endif
#else
  #error \"unknown architecture for universal build.\"
#endif
```
</details>

<details>
<summary>x64 and uint128</summary>

```
/* ABI: 64-bit */
#define MPD_CONFIG_64 1

#ifdef MPD_CONFIG_32
  #error "cannot use MPD_CONFIG_32 with 64-bit header."
#endif

#ifdef CONFIG_32
  #error "cannot use CONFIG_32 with 64-bit header."
#endif
```
</details>

**Comment by erlend-aasland** [comment]
> Might I propose an alternative way forward here however. I propose looking into not setting any CFLAGS for the system libmpdec and checking `MPD_CONFIG_64` and `MPD_CONFIG_32` in `_decimal` source code, given mpdecimal.h should tell you how it was built by setting those.

That sounds like a better way forward indeed. I've already pinged Christian on this issue, but let's wait and see if he also agrees.

**Comment by carlocab** [comment]
It's been a while here. Can we find a way to move this forward?

**Comment by erlend-aasland** [comment]
> It's been a while here. Can we find a way to move this forward?

For 3.11? Unfortunately not; 3.11 only received security fixes. For 3.12 and newer, the environment variables `LIBMPDEC_CFLAGS` and `LIBMPDEC_LIBS` are available if you want to customise how `_decimal` is built.

See also #115119.

**Comment by Bo98** [comment]
macOS still hardcodes `universal`: https://github.com/python/cpython/blob/2a5cdb251674ce8d9a824c102f7cd846d944cfa4/configure.ac#L4180-L4183. This cannot be overriden with `LIBMPDEC_CFLAGS` AFAIK as it unconditionally appends.

There was discussion of a solution above but it was waiting Christian's approval. I can look into a pull request if I find some free time soon.

**Comment by Bo98** [comment]
Looks like #115406 might have fixed this by moving the assumptions into a `AS_VAR_IF([with_system_libmpdec], [no]`. Haven't checked what branches it is in.

**Comment by erlend-aasland** [comment]
Right, so my gut feel would be that a backport of #115406 probably won't happen; such changes to the `configure` script are always slight behavioural changes to the build system, so backporting them may create havoc for distros.
